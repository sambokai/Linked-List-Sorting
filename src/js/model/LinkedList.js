class Node {
  constructor(value){
    this.value = value;
    this.next = null;
  }
}

class LinkedList {
  constructor(value) {
    this.length = 0;
    this.head = null;

    if (value) {this.append(value)}
  }

  append(value) {
    const node = new Node(value);
    let currentNode = this.head;

    if (!currentNode) {
      this.head = node;
      this.length++;

      return this;
    }

    while (currentNode.next) {
      currentNode = currentNode.next;
    }

    currentNode.next = node;

    this.length++;

    return this;
  }

  prepend(value) {
    const node = new Node(value);
    node.next = this.head;
    this.head = node;
    this.length++;
  }

  get(index){
    let current = this.head;
    let count = 0;

    if (this.length === 0) {
      throw new Error('Empty list')
    }

    if (index < 0 || index > this.length) {
      throw new Error('Invalid index')
    }

    while (count < index) {
      current = current.next;
      count++;
    }

    return current;
  }

  remove(index) {
    let current = this.head;

    if (index < 0 || index > this.length) {
      throw new Error('Invalid index')
    }

    if (index === 0) {
      this.head = current.next;
      this.length--;
      return current;
    }

    let count = 0;
    let previousNode;
    let nodeToBeDeleted;

    while (count < index) {
      previousNode = current;
      nodeToBeDeleted = current.next;
      count++;
    }

    previousNode.next = nodeToBeDeleted.next;
    this.length--;
    return nodeToBeDeleted;
  }

  clear(){
    this.head = null;
    this.length = 0;
  }

  *[Symbol.iterator](){
    let element = this.head;

    while (element) {
      yield element.value;
      element = element.next;
    }
  }
}

export default LinkedList;
