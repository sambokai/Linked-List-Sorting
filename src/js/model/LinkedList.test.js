import LinkedList from './LinkedList';


describe('LinkedList', () => {
  it('should allow initialization of a linked list by chaining append methods ', () => {
    const exampleList = new LinkedList("One").append("Two").append("Three").append("Four");
  });

  describe('prepend()', () => {
    it('should  prepend a linked list with a value', () => {
      const exampleList = new LinkedList("Two").append("Three");
      exampleList.prepend("One");

      expect(exampleList.length).toBe(3);
      expect(exampleList.head.value).toBe("One");
    });
  });

  describe('get()', () => {
    it('should return the node, located at the desired index', () => {
      const exampleList = new LinkedList("One").append("Two").append("Three").append("Four");

      expect(
        exampleList.get(0).value
      ).toBe("One");

      expect(
        exampleList.get(3).value
      ).toBe("Four");
    });

    it('should throw an error if the searched index does not exist', () => {
      const exampleList = new LinkedList("One").append("Two").append("Three").append("Four");

      const invalidTestIndexes = [-1, 20];

      invalidTestIndexes.forEach((index) =>
        expect(() =>
          exampleList.get(index)
        ).toThrowError('Invalid index'))
    });

    it('should throw an error if trying to get an element from an empty list', () => {
      const emptyList = new LinkedList();

      expect(() =>
        emptyList.get(1)
      ).toThrowError('Empty list')
    });
  });

  describe('Generator function', () => {
    it("should alllow iteration over the list's elements", () => {
      const exampleList = new LinkedList("One").append("Two").append("Three").append("Four");

      const expectedValues = ["One", "Two", "Three", "Four"];

      expect([...exampleList]).toMatchObject(expectedValues)
    });
  });
});
