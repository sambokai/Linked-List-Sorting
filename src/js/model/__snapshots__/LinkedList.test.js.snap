// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkedList clear() should delete the entire list 1`] = `
LinkedList {
  "head": null,
  "length": 0,
}
`;

exports[`LinkedList remove() should remove an element at a specified index other than 0 1`] = `
LinkedList {
  "head": Node {
    "next": Node {
      "next": Node {
        "next": null,
        "value": "Four",
      },
      "value": "Three",
    },
    "value": "One",
  },
  "length": 3,
}
`;

exports[`LinkedList remove() should remove and return an element at a specified index other than 0 1`] = `
LinkedList {
  "head": Node {
    "next": Node {
      "next": Node {
        "next": null,
        "value": "Four",
      },
      "value": "Three",
    },
    "value": "One",
  },
  "length": 3,
}
`;

exports[`LinkedList remove() should remove the first element if desired 1`] = `
LinkedList {
  "head": Node {
    "next": Node {
      "next": Node {
        "next": null,
        "value": "Four",
      },
      "value": "Three",
    },
    "value": "Two",
  },
  "length": 3,
}
`;

exports[`LinkedList remove() should remove the first element if the index is 0 1`] = `
LinkedList {
  "head": Node {
    "next": Node {
      "next": Node {
        "next": null,
        "value": "Four",
      },
      "value": "Three",
    },
    "value": "Two",
  },
  "length": 3,
}
`;

exports[`LinkedList should allow initialization of a linked list by chaining append methods  1`] = `
LinkedList {
  "head": Node {
    "next": Node {
      "next": Node {
        "next": Node {
          "next": null,
          "value": "Four",
        },
        "value": "Three",
      },
      "value": "Two",
    },
    "value": "One",
  },
  "length": 4,
}
`;
